File: 0853-car-fleet/0853-car-fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int n = position.size();
        vector<pair<int, int>> cars;

        for (int i = 0; i < n; i++) {
            cars.push_back({position[i], speed[i]});
        }

        sort(cars.rbegin(), cars.rend());

        vector<double> stack;

        for (const auto& car : cars) {
            int p = car.first;
            int s = car.second;

            double time = static_cast<double>(target - p) / s;

            if (!stack.empty() && time <= stack.back()) {
                continue;
            }

            stack.push_back(time);
        }

        return stack.size();
    }
};

File: 0853-car-fleet/0853-car-fleet.py
## time: o(nlogn)
## space : o(n)

class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        
        ##create a pair of both the position and speed arrays
        
        pair = [(p, s) for p, s in zip(position, speed)]
        
        ##create an empty stack
        stack = []
        
        ## iterate the pair in reverse order after sorting it
        for p, s in sorted(pair)[::-1]:
            
            ## calculate the distance for each car and append it to the stack
            stack.append((target - p) / s)
            
            ## check if there are two elements in the stack because lesser distance will collide with another so there must be two elements in the stack.
            
            ## check if the 2nd distance is greater than the first distance 
            
            if(len(stack) >= 2 and stack[-1] <= stack[-2]):
                stack.pop()
                
        return len(stack)       